library/core/resource_loader/src/main/java/org/quiltmc/qsl/resource/loader/impl/QuiltBuiltinResourcePackProfile.java

```diff
@@ -0,0 +1,134 @@
+
+import java.io.IOException;
+
+@ApiStatus.Internal
+public final class QuiltBuiltinResourcePackProfile extends ResourcePackProfile {
+	private static final Logger LOGGER = LogUtils.getLogger();
+
+	static QuiltBuiltinResourcePackProfile of(ModNioResourcePack pack) {
+		try {
+			PackResourceMetadata metadata = pack.parseMetadata(PackResourceMetadata.READER);
+			if (metadata == null) {
+				LOGGER.warn("Couldn't find pack meta for pack {}", pack.getName());
+				return null;
+			}
+
+			return new QuiltBuiltinResourcePackProfile(pack, metadata);
+		} catch (IOException e) {
+			LOGGER.warn("Couldn't get pack info for: {}", e.toString());
+			return null;
+		}
+	}
+
+	private QuiltBuiltinResourcePackProfile(ModNioResourcePack pack, PackResourceMetadata metadata) {
+		super(
+				pack.getName(),
+				pack.getDisplayName(),
+				pack.getActivationType() == ResourcePackActivationType.ALWAYS_ENABLED,
+				() -> pack,
+				metadata,
+				pack.type,
+				ResourcePackProfile.InsertionPosition.TOP,
+				new BuiltinResourcePackSource(pack)
+		);
+	}
+
+	@Override
+	public ResourcePackCompatibility getCompatibility() {
+		// This is to ease multi-version mods whose built-in packs actually work across versions.
+		return ResourcePackCompatibility.COMPATIBLE;
+	}
+
+	/**
+	 * Represents a built-in resource pack source.
+	 * Similar to {@link ResourcePackSource#PACK_SOURCE_BUILTIN} but specifies the mod name too.
+	 */
+	public static class BuiltinResourcePackSource implements ResourcePackSource {
+		private static final Text SOURCE_BUILTIN_TEXT = new TranslatableText("pack.source.builtin");
+		private final ModNioResourcePack pack;
+		private final Text text;
+
+		BuiltinResourcePackSource(ModNioResourcePack pack) {
+			this.pack = pack;
+
+			String displayName = pack.getDisplayName().getString();
+			String name = pack.modInfo.getName();
+			Text text = null;
+
+			// We search if the pack mentions the name of the mod from which it's coming from to determine the source text.
+			if (name != null) {
+				if (displayName.contains(name)) {
+					text = SOURCE_BUILTIN_TEXT;
+				}
+			}
+			if (displayName.contains(pack.modInfo.getId())) {
+				text = SOURCE_BUILTIN_TEXT;
+			}
+
+			// If the name doesn't say anything about the source mod, then say it in the source text.
+			// Abbreviations are needed due to GUI constraints on client.
+			if (text == null) {
+				if (name == null || (name.length() > 22 && name.length() > pack.modInfo.getId().length())) {
+					name = pack.modInfo.getId();
+				}
+
+				name = StringUtils.abbreviate(name, 22);
+				this.text = new TranslatableText("options.generic_value", SOURCE_BUILTIN_TEXT, name);
+			} else {
+				this.text = text;
+			}
+		}
+
+		@Override
+		public Text decorate(Text description) {
+			Text sourceText = this.text;
+			boolean isDescriptionSameAsRawName = description.getString().equals(this.pack.getName());
+
+			if (FabricLoader.getInstance().getEnvironmentType() == EnvType.CLIENT
```

Comment:
why truncate only on a physical client?

Reply:
Originally because the space-issue only happens on client, but the other conditions are sufficient.